halo2/src/recursion/endoscale/primitive.rs

```diff
@@ -0,0 +1,161 @@
+//! Primitives used in endoscaling.
+
+use group::{Curve, Group};
+use pasta_curves::arithmetic::{CurveAffine, FieldExt};
+
+use subtle::CtOption;
+
+/// Maps a pair of bits to a multiple of a scalar using endoscaling.
+pub(crate) fn endoscale_pair_scalar<F: FieldExt>(bits: [bool; 2]) -> F {
+    // [2 * bits.0 - 1]
+    let mut scalar = F::from(bits[0]).double() - F::one();
+
+    if bits[1] {
+        scalar *= F::ZETA;
+    }
+
+    scalar
+}
+
+/// Maps a K-bit bitstring to a scalar.
+///
+/// This corresponds to Algorithm 1 from [BGH2019], where `F` corresponds to $F_q$, the
+/// scalar field of $P$. Where Algorithm 1 computes $Acc = [scalar] P$, this function
+/// computes `scalar`.
+///
+/// [BGH2019]: https://eprint.iacr.org/2019/1021.pdf
+#[allow(dead_code)]
+pub(crate) fn endoscale_scalar<F: FieldExt>(acc: Option<F>, bits: &[bool]) -> F {
+    assert_eq!(bits.len() % 2, 0);
+
+    let mut acc = if let Some(acc) = acc {
+        acc
+    } else {
+        (F::ZETA + F::one()).double()
+    };
+    for j in (0..(bits.len() / 2)).rev() {
+        let pair = [bits[2 * j], bits[2 * j + 1]];
+        let endo = endoscale_pair_scalar::<F>(pair);
+        acc = acc.double();
+        acc += endo;
+    }
+    acc
+}
+
+/// Maps a pair of bits to a multiple of a base using endoscaling.
+pub(crate) fn endoscale_pair<C: CurveAffine>(bits: [bool; 2], base: C) -> CtOption<C> {
+    let mut base = {
+        let base = base.coordinates();
+        (*base.unwrap().x(), *base.unwrap().y())
+    };
+
+    if !bits[0] {
+        base.1 = -base.1;
+    }
+
+    if bits[1] {
+        base.0 *= C::Base::ZETA;
+    }
+
+    C::from_xy(base.0, base.1)
+}
+
+/// Maps a K-bit bitstring to a multiple of a given base.
+///
+/// This is Algorithm 1 from [BGH2019](https://eprint.iacr.org/2019/1021.pdf).
+#[allow(dead_code)]
+pub(crate) fn endoscale<C: CurveAffine>(bits: &[bool], base: C) -> C {
+    assert_eq!(bits.len() % 2, 0);
+
+    // Initialise accumulator to [2](φ(P) + P)
+    let mut acc = (base.to_curve() + base * C::Scalar::ZETA).double();
+
+    for j in (0..(bits.len() / 2)).rev() {
+        let pair = [bits[2 * j], bits[2 * j + 1]];
+        let endo = endoscale_pair::<C>(pair, base);
+        acc = acc.double();
+        acc += endo.unwrap();
+    }
+
+    acc.to_affine()
+}
+
+#[allow(dead_code)]
+pub(crate) fn i2lebsp<const NUM_BITS: usize>(int: u64) -> [bool; NUM_BITS] {
```

Comment:
Maybe this can go away once this is in `halo2_gadgets`? But we might also be able to simplify this now that we are on Rust 2021.